### DISCUSSION TITLE
Why is TaskRepository defined in service layer and not in domain layer?

### DISCUSSION THREAD

**Discussion by JanRuettinger** [MarioCarrion/todo-api-microservice-example#268] [Q&A] [answered]
You define the TasRepository [here](https://github.com/MarioCarrion/todo-api-microservice-example/blob/3e18352dcace90480b561f8b46908dc469980688/internal/service/task.go#L18) in the service layer. 

**I am curious why did you make that decision?**

I am new to DDD but I think it's a better idea to put the interface of the repository in the Domain layer so that business logic can be built on top of the functions defined in the repository.

Here is a quick example. Imagine a user is only allowed to create 5 tasks per day. That's a business rule which should be enforced in the Domain layer. In order for the Domain Model (Task) to check if 5 tasks were already created it has to make a look up in the database. This functionality can be provided by the repository.

```Go
type Task struct {
 Description string
}

func (t *Task) NewTask(ctx context.Context, description string) (Task, error) {
  if assertTaskLimit(ctx, date.Now()) != nil {
    return Task{}, error.Error("task limit reached already"
  }

  return Task{Description: description}, nil
}

type TaskRepository interface{
  assertTaskLimit(ctx context.Context, date datettime) error
}


```

Here is another [link](https://stackoverflow.com/questions/3499119/which-layer-should-repositories-go-in) to stack overflow about where to put the repository. Most answers suggest to put the interface of the repository in the domain layer.

**Comment by MarioCarrion** [comment, marked as answer]
The code you're suggesting won't work because the method in the interface is unexported, see [this playground](https://go.dev/play/p/YoD0TnF0ApP); to make it work make that method exported by changing `assertTaskLimit` to `AssertTaskLimit`; also notice the new change where `User` is the one owning tasks not `Task` so this method `(t *Task) NewTask` aligns with the problem you described originally.

With that being said what you're describing is a `User` rule not a `Task` rule; yes `Task` may have rules as well, like no empty `Task.Description` for example, but it's the `User` type that owns "no more than 5 tasks a day".

So how can we enforce that rule? One way would be to define a `User.Validate` method that does that, it makes sure the tasks for the today do not exceed 5, it literally does that, no repository calls no nothing; see [this playground](https://go.dev/play/p/d505DPIGzCX) for the implementation.

However keep in mind that although `domain.User` validates the rule we also need to add enforcement of those in the `repository.User` type to make sure no multiple customers calling at the same time for the same user add more than five: table locking, trigger or something along those lines.

_edit_: typo
